Add support arduino portenta h7

benjaminbjornsson opened the pull request.

benjaminbjornsson committed:
dts: stm32h7: Add OTG FS to DTS of the STM32H7 series

This has been verified running samples usb/console and
usb/cdc_acm on a STM32H747XIH.

Signed-off-by: Benjamin Björnsson <benjamin.bjornsson@gmail.com>

benjaminbjornsson committed:
boards: arm: arduino_portenta_h7: Add support for Arduino Portenta H7

This commit adds support for the Arduino Portenta H7 board.
The board functionallity has been verified by running following samples:
- hello_world on m7 core
- usb/console on m7 core
- blinky on m4 core

Signed-off-by: Benjamin Björnsson <benjamin.bjornsson@gmail.com>

erwango commented on boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.dts:
```diff
@@ -0,0 +1,81 @@
+	/* HW resources are split between CM7 and CM4 */
+	chosen {
+		zephyr,console = &usart1;
+		zephyr,bt-uart = &usart1;
```
?

benjaminbjornsson replied:
I'm not sure what you're referring to?

erwango replied:
`zephyr,bt-uart = &usart1;` That's a BLE specific property

erwango commented on boards/arm/arduino_portenta_h7/doc/index.rst:
```diff
@@ -0,0 +1,152 @@
+Flashing an application to STM32H747I M7 Core
+---------------------------------------------
+
+First, connect the Arduino Portenta H7 board to your host computer using
+the USB port to prepare it for flashing. Then build and flash your application.
```
Please provide more detailed instructions.
As this is DFU this is not as direct as flashing using openocd/ST-Link.

benjaminbjornsson replied:
Should of course add instructions about double tapping before flashing. But other than that I'm just using west flash as always. Should I add instructions about installing dfu-utils or just mention that it's required?

erwango replied:
> Should I add instructions about installing dfu-utils or just mention that it's required?

Not required. As dfu is already supported by west, one should be able to find dfu instructions as part of west doc (at least I hope so).

erwango commented on boards/arm/arduino_portenta_h7/doc/index.rst:
```diff
@@ -0,0 +1,152 @@
+Debugging
+=========
+
+You can debug an application in the usual way.  Here is an example for the
+:ref:`hello_world` application.
```
Is it accurate?

benjaminbjornsson replied:
No that should not be there, thanks!

erwango commented on boards/arm/arduino_portenta_h7/Kconfig.defconfig:
```diff
@@ -0,0 +1,25 @@
+
+endif # NETWORKING
+
+config DISK_DRIVER_SDMMC
+	default y if DISK_DRIVERS
```
Please remove as not enabled in dts

erwango reviewed (changes requested):
Thanks for this work.
This looks globally fine, but some clean up required

benjaminbjornsson commented on boards/arm/arduino_portenta_h7/board.c:
```diff
@@ -0,0 +1,23 @@
+
+static int board_init(const struct device *dev)
+{
+	const struct gpio_dt_spec led1 = GPIO_DT_SPEC_GET(DT_ALIAS(led1), gpios);
+
+	ARG_UNUSED(dev);
+
+	if (!device_is_ready(led1.port)) {
+		return -ENODEV;
+	}
+
+	return gpio_pin_configure_dt(&led1, GPIO_OUTPUT_INACTIVE);
```
I've chosen to turn led1 off here since I've experienced that the Arduino bootloader leaves it on.
Arduino does the same in the variantInit function found in variants.cpp.

erwango replied:
Ok, then please add a comment about it.

benjaminbjornsson commented on boards/arm/arduino_portenta_h7/Kconfig.board:
```diff
@@ -0,0 +1,15 @@
+	bool "Arduino Portenta H7 Board"
+	depends on SOC_STM32H747XX
+	select CPU_CORTEX_M7
+	select USE_STM32_LL_CRS
```
Shouldn't be there, thanks!

benjaminbjornsson commented on boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi:
```diff
@@ -0,0 +1,190 @@
+	};
+};
+
+&clk_lse {
+	status = "okay";
+};
```
Good question, what would make it required?
It should maybe be disabled by default?

erwango replied:
> Good question, what would make it required?

It would be required if used as a source by an IP (but his is not possible yet).

> It should maybe be disabled by default?

It's already disabled by default in *.dtsi

romainreignier commented:
Thnaks @benjaminbjornsson for this work!
I have started to do the same last month, but because I am new to Zephyr, the path was a bit difficult for me.

romainreignier commented on boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.dts:
```diff
@@ -0,0 +1,99 @@
+
+&rcc {
+	clocks = <&pll>;
+	clock-frequency = <DT_FREQ_M(100)>;
```
Is there a reason to not set the clock to its maximal value of 480 MHz?

benjaminbjornsson replied:
I took a look at this again which raised an issue. When trying to run the board at a higher frequency the UART gets corrupted.

My guess is that it has to do with the 25 MHz oscillator and that it's somehow inactive? I added the oscen regulator to the DTS to pull the OSCEN high during boot but is this enough?
I also just noted that in the Arduino source code they write to the PMIC over I2C to activate the 3V1 rail, which is the voltage source for the oscillator. I have tried to write to the same address in another SYS_INIT but the UART is still corrupted at higher frequencies...

romainreignier commented on boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi:
```diff
@@ -0,0 +1,125 @@
+	status = "okay";
+};
+
+zephyr_udc0: &usbotg_fs {
```
I am not very confident, but it seems to me that only the [stm32h743.dtsi](https://github.com/zephyrproject-rtos/zephyr/blob/7fe5ab315a4453322780e5655364fe4de2a1b9b6/dts/arm/st/h7/stm32h743.dtsi#L22) file has a definition for the `usbotg_fs`.

On the Portenta, as I have understood the schematic, only the High Speed interface is available without modifying R44 and R45 on the board:
![image](https://user-images.githubusercontent.com/8845664/166299218-af38c9c2-bc7e-4944-a140-511ad303def4.png)

That's why I have opened this issue: https://github.com/zephyrproject-rtos/zephyr/issues/45094

benjaminbjornsson replied:
To my understanding, the USB HS is the one connected to the boards USB-C connector. This goes through an external PHY which is not yet supported in Zephyr. The USB FS on the other hand is reachable on the Portenta Breakout board, which is what I used to verify the usb/console sample.

romainreignier replied:
Ok, I see.
Is it ok to define it here while only available with an extension board?

romainreignier replied:
I have tried the usb/console sample on the MachineControl (which is using USB FS like the Breakout board) but the ACM port does not show up on Linux.

benjaminbjornsson replied:
Have you looked at the output from the UART while running this sample? I've successfully been running both subsys/usb/cdc_acm and subsys/usb/console on my Mac.

romainreignier replied:
Sorry, I have misunderstood the MachineControl schematic, only the USB A socket is USB FS, the micro-usb is USB HS.
I could run the subsys/usb/cdc_acm and subsys/usb/console samples on my side, sorry about that.

romainreignier commented on boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.yaml:
```diff
@@ -0,0 +1,17 @@
+supported:
+  - arduino_gpio
+  - gpio
+  - arduino_spi
```
I am not sure what this line means but maybe an `arduino_i2c` could also be added?

And, should not these lines added in the dts?
```
arduino_spi: &spi2 {};
arduino_i2c: &i2c3 {};
arduino_serial: &usart1 {};
```

erwango replied:
This should be removed as arduino Uno connector is not available.
Same for `arduino_gpio`

romainreignier commented on dts/arm/st/h7/stm32h7.dtsi:
```diff
@@ -892,6 +892,26 @@
 			label = "QUADSPI";
 		};
+
+		usbotg_fs: usb@40080000 {
```
Maybe the definition in `[stm32h743.dtsi](https://github.com/zephyrproject-rtos/zephyr/blob/7fe5ab315a4453322780e5655364fe4de2a1b9b6/dts/arm/st/h7/stm32h743.dtsi)` should be removed then?
https://github.com/zephyrproject-rtos/zephyr/blob/7fe5ab315a4453322780e5655364fe4de2a1b9b6/dts/arm/st/h7/stm32h743.dtsi#L22-L34

I have noticed that the interrupts are not in the same order and h743 version has only 3 interrupts.

erwango replied:
@romainreignier Thanks for raising this.
Actually this change is not correct. Only stm32H742 "and higher" support usbotg_fs. On the other hand usbotg_hs is indeed supported on the whole stm32h7 series.

benjaminbjornsson replied:
Great catch! I'll move it to stm32h745.dtsi

romainreignier replied:
I am still unable to use the sample `samples/subsys/usb/console` on the MachineControl, the USB device is not detected by the kernel on boot.

romainreignier commented on boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi:
```diff
@@ -0,0 +1,125 @@
+		compatible = "gpio-keys";
+		wake_up: button {
+			label = "wakeup";
+			gpios = <&gpioc 13 GPIO_ACTIVE_HIGH>;
```
Which button is referred here?
I have checked the schematics and for me, the button on the Portenta is directly connected to `NRST` and the one on the Breakout Board is connected to the PMIC `PWRON`.

benjaminbjornsson replied:
You're right, this should be removed.

romainreignier commented:
I have tried your branch myself and managed to run some samples:

- basic/blinky on M7
- basic/blinky on **M4**
- hello_world on M7
- subsys/console/echo on M7
- subsys/shell/shell_module on M7
- drivers/adc on M7

For the adc sample, I had to add these lines to the common dtsi:
```
&adc1 {
	pinctrl-0 = <&adc1_inp0_pa0>;
	pinctrl-names = "default";
	status = "okay";
};
```
And this overlay in the sample:
```
/ {
	zephyr,user {
		io-channels = <&adc1 0>;
	};
};
```

I don't own the Breakout Board to try the USB FS but I have a MachineControl on which I might test it.
I also have the Ethernet Vision shield on which I plan to test the Ethernet.

Everything went well :+1:

erwango commented on boards/arm/arduino_portenta_h7/arduino_portenta_h7_m4.yaml:
```diff
@@ -0,0 +1,18 @@
+ram: 288
+flash: 1024
+supported:
+  - arduino_gpio
```
`arduino_gpio` represents the usual arduino uno connector which is not present here. Please remove.

romainreignier replied:
For the longer term, a arduino_mkr connecter device tree binding could be created but not needed for this PR.

erwango commented on boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7_defconfig:
```diff
@@ -0,0 +1,37 @@
+CONFIG_PINCTRL=y
+
+# Use zephyr,code-partition as flash offset
+CONFIG_USE_DT_CODE_PARTITION=y
```
Uncommon, is that made on purpose ?

romainreignier replied:
From my understanding of [this doc](https://docs.zephyrproject.org/latest/develop/flash_debug/host-tools.html#sam-boot-assistant-sam-ba) (*Without ROM bootloader" tab), this board has a bootloader at the start of the flash so this config allows to set the partition table in the device tree.

erwango commented on boards/arm/arduino_portenta_h7/board.c:
```diff
@@ -0,0 +1,23 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
+#include <drivers/gpio.h>
```
```suggestion
#include <zephyr/init.h>
#include <zephyr/drivers/gpio.h>
```
Cf https://github.com/zephyrproject-rtos/zephyr/pull/45417

erwango commented on boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7_defconfig:
```diff
@@ -0,0 +1,37 @@
+# SPDX-License-Identifier: Apache-2.0
```
Please add copyright

erwango reviewed (changes requested):
Few last comment on my side.

Also, can you comment on the core frequency question ?

romainreignier commented on boards/arm/arduino_portenta_h7/doc/index.rst:
```diff
@@ -0,0 +1,145 @@
+Additionally, the board features:
+- USB OTG FS
+- 3 color user LEDs
+- 1 user and reset push-button
```
There is no user button on this board

erwango replied:
@benjaminbjornsson Can you check this ?

romainreignier commented:
If I recap, the missing points on this PR are:
* A copyright
* Understand the core frequency issue
Is that all?

benjaminbjornsson commented:
> If I recap, the missing points on this PR are:
> 
> * A copyright
> * Understand the core frequency issue
>   Is that all?

Yes, I fixed the other comments locally and since I couldn’t resolve the frequency problem I didn’t push it. I’d like to solve it but I spent too many hours without progress on this…

romainreignier commented:
Ok, thanks for the update and the hard work! :+1: 
I might try as well.
Is there a mean to see at which frequency the processor is actually running?

From what I have read in the ArduinoCore code, the PMIC is configured by the bootloader. In the Arduino init code, they only reduce the current of the 3.1 V rail back to 1.2 A instead of 2 A used for the Wifi startup. It should not hurt to let it to 2A for now.

Also, from [this forum thread](https://forum.arduino.cc/t/portenta-h7-stm32cubeide-please-join-me/1010091), a guy managed to use its Portenta from CubeMX, even without the bootloader so we might some info about the clocks from this project.

romainreignier commented:
Indeed, I have tried the `hello_world` example with this value and it does not work:
```
&pll {
	div-m = <5>;
	mul-n = <192>;
	div-p = <2>;
	div-q = <2>;
	div-r = <2>;
	clocks = <&clk_hse>;
	status = "okay";
};

&rcc {
	clocks = <&pll>;
	clock-frequency = <DT_FREQ_M(480)>;
};
```

But interestingly, when running the blinky sample, the pulse duration is 4,80 seconds instead of 1 second. So there is a serious clock issue in this configuration.

![capture_blink_arduino](https://user-images.githubusercontent.com/8845664/179509378-41129db7-dce2-4514-b930-549b8ea1ec9c.PNG)

romainreignier commented:
I have tried to mimic this part of the ArduinoCore code:
https://github.com/arduino/mbed-os/blob/fb9e38d41f947ac9844dd877a1b8e613bbea1eec/targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7/system_clock_override.c#L132-L141

with:
```c
	LL_AHB4_GRP1_EnableClock(LL_AHB4_GRP1_PERIPH_GPIOH);
	LL_GPIO_SetPinMode(GPIOH, LL_GPIO_PIN_1, LL_GPIO_MODE_OUTPUT);
	LL_GPIO_SetPinOutputType(GPIOH, LL_GPIO_PIN_1, LL_GPIO_OUTPUT_PUSHPULL);
	LL_GPIO_SetPinSpeed(GPIOH, LL_GPIO_PIN_1, LL_GPIO_SPEED_FREQ_LOW);
	LL_GPIO_SetPinPull(GPIOH, LL_GPIO_PIN_1, LL_GPIO_PULL_UP);
	int wait_loop_counter = 1000;
	while(wait_loop_counter != 0)
	{
		wait_loop_counter--;
	}
	LL_GPIO_SetOutputPin(GPIOH, LL_GPIO_PIN_1);
```
Here: https://github.com/zephyrproject-rtos/zephyr/blob/d2a792dd39e588b41a0834db066b7ba387122132/drivers/clock_control/clock_stm32_ll_h7.c#L727

But nothing has changed.

So the OSCEN part should be ok. It must be a config issue somewhere.

@erwango can you spot something in the dts files?

romainreignier commented:
I have noticed that the following line might be needed in `arduino_portenta_h7_m7_defconfig`:
```
CONFIG_POWER_SUPPLY_SMPS_1V8_SUPPLIES_LDO=y
```
But it does not help by itself :(

benjaminbjornsson commented:
@romainreignier thanks for testing on your side! I’m starting to think that the best way forward may be to let the board run on the internal clock and let a future PR solve the clock issue?

romainreignier commented:
Yes, if it is acceptable by the maintainers, we could stay with this clock config for now. But make it clear in the documentation.

romainreignier commented:
Great, so now, the device tree needs to be modified to use HSI, right?

galak commented on boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.dts:
```diff
@@ -0,0 +1,98 @@
+
+	oscen: oscen {
+		compatible = "regulator-fixed-sync", "regulator-fixed";
+		label = "oscen";
```
remove label property.

galak commented on dts/arm/st/h7/stm32h745.dtsi:
```diff
@@ -35,6 +35,20 @@
+			clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x08000000>;
+			phys = <&otghs_fs_phy>;
+			status = "disabled";
+			label= "OTG_FS";
```
remove `label` property.

galak commented on dts/arm/st/h7/stm32h745.dtsi:
```diff
@@ -75,4 +89,10 @@
+	otghs_fs_phy: otghs_fs_phy {
+		compatible = "usb-nop-xceiv";
+		#phy-cells = <0>;
+		label = "OTGHS_FS_PHY";
```
remove `label` property.

galak reviewed (dismissed):
remove `label` properties that got commented on from devicetree.

benjaminbjornsson commented:
I'll try to have this PR updated this weekend.

erwango commented:
@benjaminbjornsson can you check the build issues reported by CI ?

benjaminbjornsson committed:
samples: flash_shell: Add board to platform_exclude list

Exclude Arduino Portenta H7 because the flash driver isn't
supported yet on the M4 core.

Signed-off-by: Benjamin Björnsson <benjamin.bjornsson@gmail.com>

benjaminbjornsson commented:
> @benjaminbjornsson can you check the build issues reported by CI ?

@erwango I'm confused, it's not the board being added that fails now. Should more boards be excluded from `flash_shell` or did I break something by adding the new board to the exclude list?

erwango commented:
> > @benjaminbjornsson can you check the build issues reported by CI ?
> 
> @erwango I'm confused, it's not the board being added that fails now. Should more boards be excluded from `flash_shell` or did I break something by adding the new board to the exclude list?

ok, please rebase and force push

erwango commented:
@benjaminbjornsson It seems that you're out of luck and hit be several issues non related with your commit.
2 of them were unknown till now, and I raised dedicated issues (see https://github.com/zephyrproject-rtos/zephyr/issues/48911 and https://github.com/zephyrproject-rtos/zephyr/issues/48912).
Other one is known (https://github.com/zephyrproject-rtos/zephyr/issues/48608) but seems to be a tricky one.

Meanwhile, can you check following remaning comment: https://github.com/zephyrproject-rtos/zephyr/pull/45221#discussion_r923043484 ?

romainreignier commented:
It seems that the only remaining issue is the line about the user button in the description text file.
It should not take too much time to remove @benjaminbjornsson 
Thank you.

erwango commented:
> It seems that the only remaining issue is the line about the user button in the description text file. It should not take too much time to remove @benjaminbjornsson Thank you.

PR also requires a rebase.
Note that we're now in the process of releasing V3.2.0 and PR that are not bug fixes will now have to wait that the V3.2.0 is fully released before being merged.

benjaminbjornsson commented:
Sorry for the delay, I've removed the line about the push-button in the index now and also rebased.
